### ISSUE TITLE
Generate reports

### ISSUE THREAD

**Issue by gvwilson** [swcarpentry/amy#260]
We need to generate a multitude of reports, some to include in talks and grant proposals, others to help us sync with the main web site:

- [ ] people whose training is complete but don't yet have badges
- [ ] people whose training is NULL (neither complete or incomplete)
- [ ] workshops without instructors
- [ ] export workshop payment status
- [ ] workshops for which we don't have Eventbrite
- [X] workshops for which we don't have enrolment
- [X] export workshop slugs and URLs for comparison with website
- [ ] export list of who taught/helped where and when
- [X] cumulative enrolment by date
- [X] cumulative instructors by date
- [X] cumulative workshops by date
- [ ] list instructors by frequency of teaching (including 0)
- [ ] list SCF members

We also need to update these to match what's currently in the `admin` repo (the format has changed to include names):

- [X] export YAML for badges
- [X] export YAML for airports

**Comment by r-gaia-cs** [comment]
> We need to generate a multitude of reports, some to include in talks

I need a talk in two weeks. =)

**Comment by gvwilson** [comment]
To generate YAML showing instructors by airport (for `site/config/airports.yml`) from the old admin database:

~~~
import sys
import sqlite3
import yaml

QUERY = """\                                                                                                                                          
select   airport.fullname, airport.latitude || ',' || airport.longitude, group_concat(person.person || '#' || person.personal || ' ' || person.family, '@')                                                                                                                                                
from     airport join facts join person                                                                                                               
on       airport.iata=facts.airport and facts.person=person.person                                                                                    
group by airport.iata;"""

if len(sys.argv) == 1:
    db_name = 'roster.db'
elif len(sys.argv) == 2:
    db_name = sys.argv[1]
else:
    assert False, 'Usage: airports-yaml.py [db_name]'

result = {}
connection = sqlite3.connect(db_name)
cursor = connection.cursor()
cursor.execute(QUERY)
result = [{"airport" : str(r[0]),
           "latlng" : str(r[1]),
           "who" : [x.split('#') for x in str(r[2]).split('@')]}
          for r in cursor.fetchall()]
cursor.close()
connection.close()

yaml.dump(result, sys.stdout)
~~~

**Comment by gvwilson** [comment]
To generate YAML for badges for `site/config/badges.yml` from the old admin database:

~~~
import sys
import sqlite3
import yaml

QUERY = """\                                                                                                                                          
select person.personal                                                                                                                                
       || ' '                                                                                                                                         
       || case when person.middle is null then '' when person.middle is not null then middle || ' ' end                                               
       || person.family,                                                                                                                              
       person.person                                                                                                                                  
from   person join awards                                                                                                                             
where  person.person=awards.person and awards.badge=?;"""

if len(sys.argv) == 1:
    db_name = 'roster.db'
elif len(sys.argv) == 2:
    db_name = sys.argv[1]
else:
    assert False, 'Usage: badges-yaml.py [db_name]'

result = {}
connection = sqlite3.connect(db_name)
cursor = connection.cursor()
for award in ['creator', 'instructor', 'organizer']:
    cursor.execute(QUERY, (award,))
    result[award] = [{"user" : str(r[1]), "name" : str(r[0])} for r in cursor.fetchall()]
cursor.close()
connection.close()

yaml.dump(result, sys.stdout)
~~~

**Comment by gvwilson** [comment]
Find all instructors for each workshop in old admin database:

~~~
select Event.event, group_concat(Person.personal || ' ' || Person.family, ', ') from Event join Person join Task where Task.person=Person.person and Task.event=Event.event and Task.task='instructor' group by Event.event order by Event.startdate;
~~~

**Comment by gvwilson** [comment]
Cumulative number of workshops over time from old admin database:

~~~
with counter(startdate, count) as                                                                                                                     
    (select startdate, count(*) from event                                                                                                            
     where startdate<=date('now')                                                                                                                     
     group by startdate)                                                                                                                              
select c1.startdate, sum(c2.count) from counter c1 join counter c2                                                                                    
on c1.startdate>=c2.startdate                                                                                                                         
group by c1.startdate                                                                                                                                 
order by c1.startdate;
~~~

**Comment by gvwilson** [comment]
Who to ask about missing enrolment figures from old admin database:

~~~
select event.event || ', ' || group_concat(person.personal || ' ' || person.family || ' <' || person.email || '>', ', ')                              
from event join task join person                                                                                                                      
on event.event=task.event and task.person=person.person and task.task='instructor'                                                                    
where event.event in                                                                                                                                  
    (select event.event from event where event.attendance is null and event.startdate < date('now'))                                                  
group by event.event;
~~~

**Comment by gvwilson** [comment]
Cumulative enrolment figures from old admin database:

~~~
with counter(startdate, count) as                                                                                                                     
    (select startdate, sum(attendance) from event                                                                                                     
     where attendance is not null and startdate<=date('now')                                                                                          
     group by startdate)                                                                                                                              
select c1.startdate, sum(c2.count) from counter c1 join counter c2                                                                                    
on c1.startdate>=c2.startdate                                                                                                                         
group by c1.startdate                                                                                                                                 
order by c1.startdate;
~~~

**Comment by gvwilson** [comment]
Workshops for which we do not have a detailed (person-by-person) enrolment list:

~~~
select eventbrite || ' ' || event || ' ' || url from event                                                                                            
where startdate<date('now')                                                                                                                           
and url is not null                                                                                                                                   
and eventbrite is not null                                                                                                                            
and event not in (select distinct event from task where task='learner');
~~~

**Comment by gvwilson** [comment]
How many times has each instructor taught (including people who've never taught)?

~~~
select count(*) || ' ' || person.personal || ' ' || person.family || ' <' || person.email || '>'                                                      
  from person join task join event                                                                                                                    
  on person.person=task.person and task.task='instructor' and task.event=event.event                                                                  
  where event.startdate <= date('now')                                                                                                                
  group by person.person                                                                                                                              
union                                                                                                                                                 
select 0 || ' ' || person.personal || ' ' || person.family || ' <' || person.email || '>'                                                             
  from person join awards                                                                                                                             
  on person.person=awards.person                                                                                                                      
  where awards.badge='instructor' and person.person not in                                                                                            
      (select distinct person from task where task='instructor');
~~~

**Comment by gvwilson** [comment]
Which instructors have not yet told us their skills?

~~~
select distinct person.personal || ' ' || person.family || ' <' || person.email || '>'                                                                
from person join awards on person.person=awards.person                                                                                                
where awards.badge='instructor'                                                                                                                       
and person.person not in (select distinct person from skills);
~~~

**Comment by gvwilson** [comment]
Which instructors' locations aren't recorded?

~~~
select person.personal || ' ' || person.family || ' <' || person.email || '>'                                                                         
from person join awards                                                                                                                               
on person.person=awards.person                                                                                                                        
where awards.badge='instructor'                                                                                                                       
and (                                                                                                                                                 
  (person.person not in (select person from facts))                                                                                                   
  or                                                                                                                                                  
  (person.person in (select person from facts where facts.airport is null))                                                                           
);
~~~

**Comment by gvwilson** [comment]
Cumulative number of instructors over time:

~~~
with counter(awarded, count) as                                                                                                                       
    (select awarded, count(*) from awards                                                                                                             
     where badge='instructor' and awarded<=date('now')                                                                                                
     group by awarded)                                                                                                                                
select c1.awarded, sum(c2.count)                                                                                                                      
from counter c1 join counter c2                                                                                                                       
on c1.awarded>=c2.awarded                                                                                                                             
group by c1.awarded                                                                                                                                   
order by c1.awarded;
~~~

**Comment by gvwilson** [comment]
Who counts as a member of the Software Carpentry Foundation?

~~~
with counter(awarded, count) as                                                                                                                       
    (select awarded, count(*) from awards                                                                                                             
     where badge='instructor' and awarded<=date('now')                                                                                                
     group by awarded)                                                                                                                                
select c1.awarded, sum(c2.count)                                                                                                                      
from counter c1 join counter c2                                                                                                                       
on c1.awarded>=c2.awarded                                                                                                                             
group by c1.awarded                                                                                                                                   
order by c1.awarded;
~~~

**Comment by pbanaszkiewicz** [comment]
I'm responsible for reports as part of my [GSoC proposal](https://github.com/numfocus/gsoc/blob/master/2015/proposals/banaszkiewicz-piotr-amy.md). I assume we'll go through these reports on today's call.

**Comment by pbanaszkiewicz** [comment]
>  workshops without instructors

This one's easy: `event.task_set.instructors.count() == 0`.

**Comment by wking** [comment]
On Fri, May 22, 2015 at 11:49:14AM -0700, Piotr Banaszkiewicz wrote:
> >  workshops without instructors
> 
> This one's easy: `event.task_set.instructors.count() == 0`.

Instead of reading all the events into Python and then hitting the
database to count instructors for each of them, you can do it all in
one query:

  Event.objects.exclude(task__role__name='instructor')

**Comment by gvwilson** [comment]
Closed by #279 and #361.
